src/mongocxx/collection.cpp

```diff
@@ -191,64 +191,105 @@ stdx::optional<result::bulk_write> collection::bulk_write(const class bulk_write
 
 cursor collection::find(view_or_value filter, const options::find& options) {
     using namespace bsoncxx;
-    builder::stream::document filter_builder;
-    int query_flags = ::MONGOC_QUERY_NONE;
-    scoped_bson_t filter_bson;
-    scoped_bson_t projection{options.projection()};
+    builder::stream::document opts_builder;
+    scoped_bson_t filter_bson{filter};
+    scoped_bson_t opts_bson;
 
-    filter_builder << "$query" << bsoncxx::types::b_document{filter};
+    // libmongoc::collection_find_with_opts does not support the legacy "modifiers" options, so we
+    // must copy the options struct and convert all of the modifiers options to their modern
+    // equivalents.
```

Comment:
I suggest pulling all the filter/opt builder code into a separate private function so that `find` itself is shorter and focused on sending the filter and getting the cursor ready.   E.g. pass filter and options and a reference to a filter_builder and opts_builder and let the function modify the builders as appropriate.